### PR TITLE
Lint `manual_unwrap_or` for it let cases

### PR DESCRIPTION
**Pull request by lochetti**
This PR modifies `manual_unwrap_or` to lint for `if let` cases as well. This effort is part of the fixes desired by https://github.com/rust-lang/rust-clippy/issues/12618

changelog:[`manual_unwrap_or`]: Lint for `if let` cases.

### CODE REVIEW

**Comment by Manishearth** [comment]
r? @y21 

might not get time to get around to this in the next few days

File: clippy_lints/src/matches/manual_unwrap_or.rs
Comment (y21) [reply clippy_lints/src/matches/manual_unwrap_or.rs:55]: ah, I now realise that this code was just moved around and isn't actually new code.

File: tests/ui/manual_unwrap_or_default.rs
Comment (lochetti) [reply tests/ui/manual_unwrap_or_default.rs:2]: @y21 we have this test, that is the one that started to emit the `manual_unwrap_or`
```    
// Issue #12564  
// No error as &Vec<_> doesn't implement std::default::Default
let mut map = std::collections::HashMap::from([(0, vec![0; 3]), (1, vec![1; 3]), (2, vec![2])]);
let x: &[_] = if let Some(x) = map.get(&0) { x } else { &[] };
```

Until before this PR it didn't emit the `manual_unwrap_or` because we didn't have `manual_unwrap_or` implemented for if let. 

But, I think that we already have this problem of emiting both lints for some cases. If we go to master and simply try this code
```
    let x: Option<i32> = None;
    match x {
        Some(v) => v,
        None => 0,
    };
```
it will emit both lints. And this is also reaffirmed by the fact that all test cases in `tests/ui/manual_unwrap_or_default.rs` are not using `i32` and `0` for instance, they are using `String` and `String::new()`, for instance.

So it look likes we already have a problem right?
If so, and in case we want to avoid linting both lints, maybe we could not lint the `manual_unwrap_or` if the constant value is the default? But what if the developer had avoided  `manual_unwrap_or_default`?

Comment (y21) [reply tests/ui/manual_unwrap_or_default.rs:2]: ...Ah, okay I see it now. I think the ideal solution would be to unify these two lints and have them live in the same file, because as far as I understand, the only difference between these two lints is the kind of expression in the `else` case ("if it's equivalent to a `default()` call, lint `manual_unwrap_or_default`, otherwise `manual_unwrap_or`"), all of the additional lint logic seems like it should be exactly the same.

But since this was already an issue before this PR and this is just a side effect of this linting more cases, it doesn't need to be done here in this PR. 

> But what if the developer had avoided `manual_unwrap_or_default`?

That's a fair point. We do have the function [`is_lint_allowed`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.is_lint_allowed.html) for checking if a lint is `#[allow]`ed. That would allow us to see if the user explicitly disabled it, and we could still emit this I think.

But as mentioned above, it doesn't really need to be done in this pr.

File: clippy_lints/src/matches/manual_unwrap_or.rs
Comment (lochetti) [reply clippy_lints/src/matches/manual_unwrap_or.rs:55]: And here, I tried to understand, but I also don't know if I need to to anything else :grimacing:

File: tests/ui/manual_unwrap_or_default.rs
Comment (lochetti) [reply tests/ui/manual_unwrap_or_default.rs:2]: Ok! Understood. 
After we merge this PR, I will create a new issue for this "merge" of both lints at the same file :)

**Review by y21 — APPROVED** [review]

**Comment by y21** [comment]
LGTM, thanks! The other two comments were issues that already existed so I don't think we really need to block this PR on that, it's a good improvement on its own. (I'll get to them if nobody else does)

@bors r+

**Pull request merged**
